File: 206/Reverse_Linked_List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

/** Recursive
 * struct ListNode* reverseList(struct ListNode* head) {
 *     if (head == NULL || head->next == NULL) return head;
 * 
 *     struct ListNode *cur = reverseList(head->next);
 *     head->next->next = head;
 *     head->next = NULL;
 *     return cur;
 * }
 */

// Iterative
struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode *cur = head, *pre = NULL, *tmp = NULL;
    while(cur) {
        tmp = cur->next;
        cur->next = pre;
        pre = cur;
        cur = tmp;
    }
    return pre;
}

File: 206/Reverse_Linked_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # recursive
        # if not head or not head.next:
        #     return head
        # cur = self.reverseList(head.next)
        # head.next.next = head
        # head.next = None
        # return cur
        
        # iterative
        pre, cur = None, head
        while cur:
            nextNode = cur.next
            cur.next = pre
            pre = cur
            cur = nextNode
        return pre
